runtime/stdlib/hashalgorithm.go

```diff
@@ -25,100 +25,97 @@ import (
 	"github.com/onflow/cadence/runtime/sema"
 )
 
-var hashAlgorithmFunctions = map[string]interpreter.FunctionValue{
-	sema.HashAlgorithmTypeHashFunctionName:        hashAlgorithmHashFunction,
-	sema.HashAlgorithmTypeHashWithTagFunctionName: hashAlgorithmHashWithTagFunction,
+var hashAlgorithmTypeID = sema.HashAlgorithmType.ID()
+var hashAlgorithmStaticType interpreter.StaticType = interpreter.CompositeStaticType{
+	QualifiedIdentifier: sema.HashAlgorithmType.Identifier,
+	TypeID:              hashAlgorithmTypeID,
 }
 
-func NewHashAlgorithmCase(inter *interpreter.Interpreter, rawValue uint8) *interpreter.CompositeValue {
-	return interpreter.NewEnumCaseValue(
-		inter,
-		interpreter.ReturnEmptyLocationRange,
-		sema.HashAlgorithmType,
-		interpreter.NewUInt8Value(inter, func() uint8 {
-			return rawValue
-		}),
-		hashAlgorithmFunctions,
+func NewHashAlgorithmCase(rawValue interpreter.UInt8Value) interpreter.MemberAccessibleValue {
+
+	value := interpreter.NewSimpleCompositeValue(
+		nil,
+		sema.HashAlgorithmType.ID(),
+		hashAlgorithmStaticType,
+		[]string{sema.EnumRawValueFieldName},
+		nil,
+		nil,
+		nil,
+		nil,
 	)
+	value.Fields = map[string]interpreter.Value{
+		sema.EnumRawValueFieldName:                    rawValue,
+		sema.HashAlgorithmTypeHashFunctionName:        hashAlgorithmHashFunction(value),
+		sema.HashAlgorithmTypeHashWithTagFunctionName: hashAlgorithmHashWithTagFunction(value),
+	}
+	return value
 }
 
-var hashAlgorithmHashFunction = interpreter.NewUnmeteredHostFunctionValue(
-	func(invocation interpreter.Invocation) interpreter.Value {
-		dataValue, ok := invocation.Arguments[0].(*interpreter.ArrayValue)
-		if !ok {
-			panic(errors.NewUnreachableError())
-		}
-		hashAlgoValue := invocation.Self
-
-		inter := invocation.Interpreter
-
-		getLocationRange := invocation.GetLocationRange
-
-		inter.ExpectType(
-			hashAlgoValue,
-			sema.HashAlgorithmType,
-			getLocationRange,
-		)
-
-		return inter.HashHandler(
-			inter,
-			getLocationRange,
-			dataValue,
-			nil,
-			hashAlgoValue,
-		)
-	},
-	sema.HashAlgorithmTypeHashFunctionType,
-)
+func hashAlgorithmHashFunction(hashAlgoValue interpreter.MemberAccessibleValue) *interpreter.HostFunctionValue {
```

Comment:
previously this function was only defined once, but now it creates a one for every enum-case of the hash algorithm. Is it no longer possible to keep the old behaviour (optimized)?

Reply:
That was actually done in #1813. I'll open a follow-up PR

Reply:
Looked into it. The problem is that #1813 refactored the enum cases from `CompositeValue`s to `SimpleCompositeValue`. For `CompositeValue`, `GetMember` returns a `BoundFunctionValue`, which sets `Invocation.Self`. But `SimpleCompositeValue` does not do this, as it internally makes no distinction between fields and functions.

We can either:
- Keep the current code as-is. The overhead of having a separate function per case should be negligible; or
- Distinguish between fields and functions in `SimpleCompositeValue` and return a `BoundFunctionValue` for functions

Thoughts?

Reply:
I still need to go through the code again to fully understand what this means / and the impact.

Meantime this is certainly not a blocker for merging the PR 👍